packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -303,26 +286,124 @@ export class OdspDocumentService implements IDocumentService {
+                    switch (code) {
+                        case "sessionForbiddenOnPreservedFiles":
+                        case "sessionForbiddenOnModerationEnabledLibrary":
+                        case "sessionForbiddenOnRequireCheckout":
+                            // This document can only be opened in storage-only mode.
+                            // DeltaManager will recognize this error
+                            // and load without a delta stream connection.
+                            this._policies = {...this._policies,storageOnly: true};
+                            throw new DeltaStreamConnectionForbiddenError(code, { driverVersion });
+                        default:
+                            continue;
+                    }
+                }
+            }
+            throw e;
+        });
+    }
+
+    private async joinSessionCore(
+        requestSocketToken: boolean,
+        options: TokenFetchOptionsEx,
     ): Promise<ISocketStorageDiscovery> {
-        const executeFetch = async () =>
-            fetchJoinSession(
-                this.odspResolvedUrl,
-                "opStream/joinSession",
-                "POST",
-                this.mc.logger,
-                this.getStorageToken,
-                this.epochTracker,
-                requestSocketToken,
-                options,
-                this.hostPolicy.sessionOptions?.unauthenticatedUserDisplayName,
-            );
+        const disableJoinSessionRefresh = this.mc.config.getBoolean("Fluid.Driver.Odsp.disableJoinSessionRefresh");
+        const executeFetch = async () => {
+            return {
+                    entryTime: Date.now(),
+                    joinSessionResponse: await fetchJoinSession(
+                        this.odspResolvedUrl,
+                        "opStream/joinSession",
+                        "POST",
+                        this.mc.logger,
+                        this.getStorageToken,
+                        this.epochTracker,
+                        requestSocketToken,
+                        options,
+                        disableJoinSessionRefresh,
+                        this.hostPolicy.sessionOptions?.unauthenticatedUserDisplayName,
+                    )
+            };
+        };
+
+        let response = await this.cache.sessionJoinCache.addOrGet(this.joinSessionKey, executeFetch);
+        // If the response does not contain refreshSessionDurationSeconds, then treat it as old flow and let the
+        // cache entry to be treated as expired after 1 hour.
+        let refreshSessionDurationSeconds: number | undefined =
+            response.joinSessionResponse.refreshSessionDurationSeconds ?? 3600;
+        if (disableJoinSessionRefresh) {
+            const expiryTime = response.entryTime + refreshSessionDurationSeconds * 1000;
+            if (expiryTime <= Date.now()) {
+                this.cache.sessionJoinCache.remove(this.joinSessionKey);
```

Comment:
I think the code might be a bit more streamlined if we stored not entryTime but expiration time in the cached entry. Then all the logic RE when/how to store it would be in one place and reading code would not need to care about disableJoinSessionRefresh fork. Note that value of disableJoinSessionRefresh might change over time (from session to session) - probably not a big deal, but something to consider

Reply:
Actually, can we completely remove if (disableJoinSessionRefresh) flow (true case)? I think your else case is covering it correctly. All you need to do is for calculateJoinSessionRefreshDelta() to take it into account, which would be a good think - code would be more streamlined.
Also, the fewer states we have in the code the better. Ideally disableJoinSessionRefresh is used only
- to patch response.joinSessionResponse.refreshSessionDurationSeconds response when we fetch it (the rest of the code does not care if value came from SPO or not).
- to set timer or not.
The rest of the code can not care....

Reply:
I removed the disableJoinSessionRefresh fork for old flow. It looks more simplified now. Also patching the refreshSessionDurationSeconds to 1 hour in case of old flow.